### DISCUSSION TITLE
Global rather then per-component parameters

### DISCUSSION THREAD

**Discussion by shalevm** [nasa/fprime#1603] [Q&A] [answered]
Hi,
I'd like to define several parameters which are global, namely, can be used by more then one component (e.g. the moon's G). I'm not really sure what is the best approach in this case, since parameters are defined per-component. 
The only solution I can think of seems wrong and cumbersome: to define a component for global parameters, which other components can communicate with via dedicated ports.
any ideas?
Thanks.

**Comment by codeflight1** [comment]
If the values are going to be constant, you can define a custom constant fpp file that will be compiled into a header that you can include in your components.

**Reply by shalevm** [reply]
thanks, but the problem is with the non constant parameters.

**Comment by SterlingPeet** [comment]
My preference for this type of situation is to designate a single component to “own” the parameter, and then other components ask for the value using a port connected to that component.   It doesn’t scale nicely, and if I needed them to be used across a lot of components the I would probably try to architect something around a central service component.

Otherwise, you can solve this problem in your ground data segment by writing some glue code to batch together to commands that would change and save multiple parameters at the same time.  You could get a good expectation of atomicity if you upload these batched commands as sequences and the command sequencer will run them all back-to-back.

Sterling

Sent from my mobile.

On Jul 31, 2022, at 10:32 AM, shalevm ***@***.***> wrote:

﻿

thanks, but the problem is with the non constant parameters.

**Reply by shalevm** [reply]
thanks, I went for the first option, but the question then becomes what is the right way "to architect something around a central service component". specifically, Is there support in the fprime framework for a notification mechanism, to let the other components know that the values in the service component have been updated? (e.g. subscribing to an event with an event handler)
Thanks.

**Comment by timcanham** [comment, marked as answer]
A possible solution that other projects have used is to have one component own it and then push it as an entry to `Svc/PolyDb`.

**Reply by shalevm** [reply]
thanks a lot Tim - that's what I was looking for. 
Is there support for a notification mechanism, to let the other components know that the values in the PolyDb have been updated and thus need to be retrieved (so components won't need to poll it)?
Thanks again.

**Comment by timcanham** [comment]
No, `PolyDb` is very simple - just a push/pull, no notifications.

**Reply by shalevm** [reply]
is there a framework component that does that? or is it just a matter of creating a one-to-many topology that enables sending a message from a port on one component to ports on several components at once (is that even allowed)?

**Comment by SterlingPeet** [comment]
F Prime supports events as a built in system.  Using an update event to trigger fetching a parameter from the PolyDb is feasible, but will require plumbing the event to a port for your secondary use components.  That bit could mean distributing an event to many components, or routing events based on contents to the correct component.

Sterling

Sent from my mobile.

On Aug 1, 2022, at 5:50 AM, shalevm ***@***.***> wrote:

﻿

is there a framework component that does that? or is it just a matter of creating a one-to-many topology that enables sending a message from a port on one component to ports on several components at once (is that even allowed)?

**Reply by shalevm** [reply]
Thanks Sterling.
one thing that's still unclear to me: to my understanding, fprime events just represent a log of activities taken by the system. so i'm not really clear as to how to use them for "plumbing the event to a port for your secondary use components".
thanks again.

**Reply by SterlingPeet** [reply]
You're welcome.  Fundamentally, everything in F Prime boils down to two primitives, Components and Ports.  There is plenty about this in the documentation.  Event Ports are a specific Port type, and the autocoder does some special things to generate them and the supporting code to move them around without the mission developer needing to implement the common parts in every component.

However, they are just typed messages that are passed between components via ports.  You can design your topology to intercept/use/pass these messages like any other ports.  

In order to understand and interpret them in your code, you may need to go look at the code (or at least the SDD document) in the ``Fw/Log`` folder where the serialized EVR log buffers are defined.  There are likely other places that you need to go investigate, but this should get you started.

Handling them is just a matter of interpreting the message, and determining what to do with it.  In the Ref app, events are all connected to the ``ActiveLogger`` component that packages them into a downlink message buffer, and sends them to the next component in the downlink handling chain [which I believe is currently the Framer component] so that the events arrive at the GDS ready to be unpacked and displayed.  It would be reasonable to write a component that accepts events via a LogIn port, inspects for certain contents, and then passes the event on to its own LogOut port to continue being handled in the normal way.  If the contents match a convenient filter parameter, it would _also_ reach out on a custom update port to each necessary component before passing on the event.

If you use this approach, then the update port would signal to a component to retrieve new parameters from the PolyDb or whatever storage component is being used.  That port doesn't even need to contain information, it could just kick off a generic parameter update within the target component.

**Reply by shalevm** [reply]
that's awesome. thanks Sterling.
